### PR TITLE
Pulumi program testing library

### PR DESCRIPTION
**Pull request by danielrbradley**
Adds an experimental new library for constructing tests which interact with the Pulumi CLI. Once stabalised and proven, this could be a candidate to be re-homed into pulumi/pulumi.

# Pulumi Testing Library

Pulumi Test is a thin(ish) wrapper over the Automation API, making it easier to use within test scenarios.

The Automation API is just a thin wrapper around the calling the Pulumi CLI (`pulumi ...`). Some additional commands which are not yet supported by the Automation API are also added for convenience.

## Getting Started

The starting point to testing a program is to create a new PulumiTest pointing at some existing test.

```go
func TestPulumiProgram(t *testing.T) {
  test := NewPulumiTest(t, filepath.Join("path", "to", "program"))
  //...
}
```

By default run your program is copied to a temporary directory before running to avoid cluttering your working directory with temporary or ephemeral files. To disable this behaviour, use `NewPulumiTestInPlace`. You can also do a copy of a test manually by calling `CopyToTempDir()`:

```go
source := NewPulumiTestInPlace(t, ...)
copy := source.CopyToTempDir()
```

The `source` variable is still pointing to the original path, but the `copy` is pointing to a new PulumiTest in temporary directory which will automatically get removed at the end of the test.

Before we can preview or deploy a program we need to install dependencies and create a stack:

```go
//...
test.Install() // Runs `pulumi install` to restore all dependencies
test.NewStack("my-stack") // Creates a new stack and returns it.
```

These two steps can also be done together by calling `InstallStack()`:

```go
test.InstallStack("my-stack")
```

The created stack is returned but is also set as the current stack on the PulumiTest object. All methods such as `source.Preview()` or `source.Up()` will use this current stack.

> [!NOTE]
> The new stack will be automatically destroyed and removed at the end of the test.

## Default Settings

`PULUMI_BACKEND_URL` is set to a temporary directory. This improves test performance and doesn't rely on the user already being authenticated to a specific backend account. This also isolates stacks so the same stack name can be re-used for several tests at once without risking conflicts and avoiding stack name randomisation which breaks importing & exporting between test runs. This can be overridden by the option `opttest.UseAmbientBackend()` or by setting `PULUMI_BACKEND_URL` yourself in the stack initialization options.

`PULUMI_CONFIG_PASSPHRASE` is set by default to "correct horse battery staple" (an arbitrary phrase) so that encrypted values are not tied to an external secret store that the user might not have access to. This can be overridden by setting `PULUMI_CONFIG_PASSPHRASE` in the stack initialization options.

## Configuring Providers

Pulumi discovers plugins the same as when running Pulumi commands directly.

In a test scenario, we often want to ensure a specific implementation of a provider is used during testing. The most reliable way is to configure use plugin attachment `PULUMI_DEBUG_PROVIDERS=NAME:PORT`. This prevents the Pulumi engine from searching for and starting the provider with the given name. Instead, it will connect to the already-running provider on the specified port. If the provider is not reachable on the given port, Pulumi will throw an error.

These can be specified via the `Attach*` options when constructing the test:

```go
// Start a provider yourself
NewPulumiTest(t, "path", opttest.AttachProvider("gcp", func(ctx context.Context) (int, error) {
  return port, nil // TODO: Actually start a provider.
})
// Start a server for testing from a pulumirpc.ResourceProviderServer implementation
NewPulumiTest(t, "path", opttest.AttachProviderServer("gcp", func() (pulumirpc.ResourceProviderServer, error) {
  return makeProvider()
})
// Specify a local path where the binary lives to be started and attached.
NewPulumiTest(t, "path", opttest.AttachProviderBinary("gcp", filepath.Join("..", "bin"))
// Use Pulumi to download a specific published version, then start and attach it.
NewPulumiTest(t, "path", opttest.AttachDownloadedPlugin("gcp", "6.61.0")
```

For providers which don't support attaching, we can configure the path to the binary of a specific provider in the `plugins.providers` property in the project settings (Pulumi.yaml) by using the `LocalProviderPath()` option:

```go
NewPulumiTest(t, "path", opttest.LocalProviderPath("gcp", filepath.Join("..", "bin"))
```

## Pulumi Operations

Preview, Up, Refresh and Destroy can be run directly from the test context:

```go
test.Preview()
test.Up()
test.Refresh()
test.Destroy()
```

> [!NOTE]
> Stacks created with `InstallStack` or `NewStack` will be automatically destroyed and removed at the end of the test.

## Using Local SDKs

When running tests via SDKs that haven't yet been published, we need to configure the program under test to use our local build of the SDK instead of installing a version from their package registry.

### Node.js (Yarn Link)

For Node.js, we support using a [locally linked](https://classic.yarnpkg.com/lang/en/docs/cli/link/) version of the NPM package. Before running your test, you must run `yarn link` in the directory of the built Node.js SDK (normally in `sdk/nodejs/bin`).

Once the local link is configured, you can configure your test to use the linked package by using the `YarnLink` test option:

```go
NewPulumiTest(t, "test_dir", opttest.YarnLink("@pulumi/azure-native"))
```

### Go - Module Replacement

In Go, we support adding a replacement to the go.mod of the program under test. This is implemented by calling [`go mod edit -replace`](https://pkg.go.dev/cmd/go#hdr-Edit_go_mod_from_tools_or_scripts) with the user-specified replacement.

The replacement can be specified using the `GoModReplacement` test option:

```go
NewPulumiTest(t, "test_dir",
  opttest.GoModReplacement("github.com/pulumi/pulumi-my-provider/sdk/v3", "..", "sdk"))
```

### Update Source

Update source files for a subsequent step in the test:

```go
test.UpdateSource("folder_with_updates")
```

### Set Config

Set a variable in the stack's config:

```go
test.SetConfig("gcp:project", "pulumi-development")
```

## Asserts

The `assertup` and `assertpreview` modules contain a selection of functions for asserting on the results of the automation API:

```go
assertup.HasNoDeletes(t, upResult)
assertup.HasNoChanges(t, upResult)
assertpreview.HasNoChanges(t, previewResult)
assertpreview.HasNoDeletes(t, previewResult)
```

## Example

Here's a complete example as a test might look for the gcp provider with a local pre-built binary.

```go
func TestExample(t *testing.T) {
  // Copy test_dir to temp directory, install deps and create "my-stack"
  test := NewPulumiTest(t, "test_dir", opttest.AttachProviderBinary("gcp", "../bin"))
  test.InstallStack("my-stack")

  // Configure the test environment & project
  test.SetConfig("gcp:project", "pulumi-development")

  // Preview, Deploy, Refresh, 
  preview := test.Preview()
  t.Log(preview.StdOut)

  deploy := test.Up()
  t.Log(deploy.StdOut)
  assertpreview.HasNoChanges(t, test.Preview())

  // Export import
  test.ImportStack(test.ExportStack())
  assertpreview.HasNoChanges(t, test.Preview())

  test.UpdateSource(filepath.Join("testdata", "step2"))
  update := test.Up()
  t.Log(update.StdOut)
}
```

Comparative ProgramTest example:

```go
func TestExample(t *testing.T) {
  test := integration.ProgramTestOptions{
    Dir: testDir(t, "test_dir"),
    Dependencies: []string{filepath.Join("..", "sdk", "python", "bin")},
    ExpectRefreshChanges: true,
    Config: map[string]string{
      "gcp:project": "pulumi-development",
    },
    LocalProviders: []integration.LocalDependency{
      {
        Package: "gcp",
        Path:    "../bin",
      },
    },
    EditDirs: []integration.EditDir{
      {
        Dir:      testDir(t, "test_dir", "step2"),
        Additive: true,
      },
    },
  }

  integration.ProgramTest(t, &test)
}
```

### CODE REVIEW

File: pulumitest/asserthelpers/helpers.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
+package asserthelpers
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/apitype"
+)
+
+func FindUnexpectedOps(changeSummary map[string]int, expectedOpTypes ...apitype.OpType) []string {
```
Comment (t0yv0) [inline pulumitest/asserthelpers/helpers.go:10]: Well this is interesting looks like Automation API exposes change results. Which is great! And we have a model for it in apitype.OpType

https://github.com/pulumi/pulumi/blob/master/sdk/go/common/apitype/history.go#L63

This is great. 

Can we hide the "string" from public API here then and use apitype.OpType everywhere?

File: pulumitest/pulumiTest.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	source       string
+	options      []opttest.Option
+	currentStack *auto.Stack
+}
+
+func NewPulumiTest(t *testing.T, source string, opts ...opttest.Option) *PulumiTest {
+	return NewPulumiTestInPlace(t, source, opts...).CopyToTempDir()
+}
+
+func NewPulumiTestInPlace(t *testing.T, source string, opts ...opttest.Option) *PulumiTest {
```
Comment (t0yv0) [inline pulumitest/pulumiTest.go:23]: Curious, any reason it's not just an option on NewPulumiTest?

File: pulumitest/pulumiTest_test.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+	"github.com/pulumi/providertest/providers"
+	"github.com/pulumi/providertest/pulumitest/assertpreview"
+	"github.com/pulumi/providertest/pulumitest/assertup"
+	"github.com/pulumi/providertest/pulumitest/opttest"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/apitype"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestDeploy(t *testing.T) {
```
Comment (t0yv0) [inline pulumitest/pulumiTest_test.go:15]: Brilliant needed test-tests badly.

File: pulumitest/preview.go [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package pulumitest
+
+import (
+	"github.com/pulumi/pulumi/sdk/v3/go/auto"
+	"github.com/pulumi/pulumi/sdk/v3/go/auto/optpreview"
+)
+
+// Preview previews an update to the current stack.
+func (a *PulumiTest) Preview(opts ...optpreview.Option) auto.PreviewResult {
```
Comment (t0yv0) [inline pulumitest/preview.go:9]: If we're doing wrappers, do wrappers all the way here.. Wrap auto.PreviewResult so we can add helper methods at the level of this library.

Comment (t0yv0) [reply pulumitest/preview.go:9]: Just an idea.

File: providers/attach.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
+package providers
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+)
+
+func StartProviders(ctx context.Context, factories map[string]ProviderFactory) (map[string]int, context.CancelFunc, error) {
```
Comment (t0yv0) [inline providers/attach.go:10]: What is the string in factories? A bit of documentation can go a long way. AS well as what is `map[string]int`? Is this like "aws"? aNd ports?

File: providers/downloadPluginBinary.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+		if err != nil {
+			return 0, err
+		}
+		return startLocalBinary(ctx, binaryPath, name)
+	}
+	return factory
+}
+
+func DownloadPluginBinary(name, version string) (string, error) {
+	cmd := exec.Command("pulumi", "plugin", "install", "resource", name, version)
```
Comment (t0yv0) [inline providers/downloadPluginBinary.go:23]: Eh this one is a bit interesting. It uses global plugin cache which is reasonable. One day perhaps allow user to seal off the local plugin cache for the tests. Can do later.

File: providers/factory.go [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package providers
+
+import "context"
+
+// ProviderFactory is a function that starts a provider and returns the port it is listening on.
+// The function should return an error if the provider fails to start.
+// When the test is complete, the context will be cancelled and the provider should exit.
+type ProviderFactory func(ctx context.Context) (int, error)
```
Comment (t0yv0) [inline providers/factory.go:8]:
```
type Port int

ype ProviderFactory func(ctx context.Context) (Port, error)
```

File: pulumitest/opttest/opttest.go [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+// GoModReplacement specifies replacements to be add to the go.mod file when running the program under test.
+// Each replacement is added to the go.mod file with `go mod edit -replace <replacement>` on stack creation.
+func GoModReplacement(packageSpecifier string, replacementPathElem ...string) Option {
+	return optionFunc(func(o *Options) {
+		o.GoModReplacements[packageSpecifier] = filepath.Join(replacementPathElem...)
+	})
+}
+
+// UseAmbientBackend configures the test to use the ambient backend rather than a local temporary directory.
```
Comment (t0yv0) [inline pulumitest/opttest/opttest.go:64]: Not sure what this means still.

File: pulumitest/newStack.go [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+		err = stack.Workspace().RemoveStack(pt.ctx, stackName, optremove.Force())
+		if err != nil {
+			pt.t.Errorf("failed to remove stack: %s", err)
+		}
+	})
+	pt.currentStack = &stack
+	return &stack
+}
+
+func randomStackName(dir string) string {
```
[the remark this answers is not in the document]
Comment (danielrbradley) [reply pulumitest/newStack.go:163]: Yup, same implementation ... extracted out from program test in case we want compatibility here.

**Comment by t0yv0** [comment]
🚢 

I think this is very nice. This is a well factored foundation that has all the features that ProgramTester.ManualLifecycle and then some improvements that we were needing badly in providertest.. It's not yet "proven" by  a lot of use but it has no legacy cruft and looks very easy to maintain. Extra props for including tests from the get-go.

The state traditions are modeled imperatively and this being Go that seems like the way.. 

I'm thinking that I'm not seeing anything missing here to implement upgrade test cleanly. One thing perhaps, this:

```
opttest.AttachProviderBinary("gcp", "../bin")
```

This is currently a 'global' option. But it shouldn't be. I want to be able to imperatively change up dependencies in the lifecycle. This is the essence of upgrade testing. Could we make it imperatively modified same way you have this:

```
test.UpdateSource(filepath.Join("testdata", "step2"))
```

We could then do this:

```
func TestPreviewOnlyUpgrade(t *testing.T) {
  test := NewPulumiTest(t, "test_dir")
  test.InstallStack("my-stack") // install baseline versions

  // Configure the test environment & project
  test.SetConfig("gcp:project", "pulumi-development")

  test.ImportStack("..saved..snapshot..") // grab from baseline versions

  test.SetDependencies(opttest.AttachProviderBinary("gcp", "../bin")) // now we're testing the local build
  test.UpdateSource(filepath.Join("testdata", "step2")) // optionally simulate the user updating the source on upgrade
  deploy := test.Preview()
  assertpreview.HasNoChanges(t, test.Preview())
}
```

And for "real" upgrade test we could do this:

```
func TestRealUpgrade(t *testing.T) {
  test := NewPulumiTest(t, "test_dir")
  test.InstallStack("my-stack") // install baseline versions

  // Configure the test environment & project
  test.SetConfig("gcp:project", "pulumi-development")

  test.Up()

  test.SetDependencies(opttest.AttachProviderBinary("gcp", "../bin")) // now we're testing the local build
  test.UpdateSource(filepath.Join("testdata", "step2")) // optionally simulate the user updating the source on upgrade
  deploy := test.Up()
  assertpreview.HasNoChanges(t, deploy)
}
```

**Review by t0yv0 — APPROVED** [review]

File: pulumitest/preview.go (hunk 4 above)
Comment (danielrbradley) [reply pulumitest/preview.go:9]: Maybe, though this increases the implementation size and adds more elements to keep in-sync with the underlying implementation. I'd delay putting this in for the return types for now and re-evaluate if we hit any messy cases as-is.

File: providers/attach.go (hunk 5 above)
Comment (danielrbradley) [reply providers/attach.go:10]: Yup ... adding better typing here 👍

File: pulumitest/opttest/opttest.go (hunk 8 above)
Comment (danielrbradley) [reply pulumitest/opttest/opttest.go:71]: Here's my next attempt:

> UseAmbientBackend skips setting PULUMI_BACKEND_URL a local temporary directory which overrides any backend configuration which might have been done on the local environment via `pulumi login`.
> Using this option will cause the program under test to use whatever backend configuration has been set via `pulumi login` or an existing `PULUMI_BACKEND_URL` value.

**Comment by blampe** [comment]
> ## Default Settings
>
> PULUMI_BACKEND_URL is set to a temporary directory. This improves test performance and doesn't rely on the user already being authenticated to a specific backend account. This also isolates stacks so the same stack name can be re-used for several tests at once without risking conflicts and avoiding stack name randomisation which breaks importing & exporting between test runs. This can be overridden by the option opttest.UseAmbientBackend() or by setting PULUMI_BACKEND_URL yourself in the stack initialization options.
>
> PULUMI_CONFIG_PASSPHRASE is set by default to "correct horse battery staple" (an arbitrary phrase) so that encrypted values are not tied to an external secret store that the user might not have access to. This can be overridden by setting PULUMI_CONFIG_PASSPHRASE in the stack initialization options.

These are exceptionally reasonable defaults. A lot of integration tests currently hit the Pulumi Cloud service, but this generates an enormous amount of clutter in the system not to mention another source of flakiness.

**Comment by danielrbradley** [comment]
Going to merge as-is so we can start integrating this and can continue to iterate on additional features in follow-up PRs.

File: pulumitest/opttest/opttest.go (hunk 8 above)
Comment (t0yv0) [reply pulumitest/opttest/opttest.go]: Ah I understand what it means now, sorry it's on me. This is about the Pulumi state backend.

**Pull request merged**